### PR TITLE
Improve lock mechanism in tso.KeyspaceGroupManager

### PR DESCRIPTION
**Pull request by binshi-bing**
### What problem does this PR solve?

Issue Number: Ref #6232 

### What is changed and how does it work?

```commit-message
Use the RWMutex instead of individual atomic types to better protect the state of the keyspace group manager
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/tso/allocator_manager.go [hunk 1]
```diff
@@ -167,8 +167,8 @@ type AllocatorManager struct {
 
 	ctx    context.Context
 	cancel context.CancelFunc
-	// ksgID is the keyspace group id
-	ksgID uint32
+	// kgID is the keyspace group id
```
Comment (JmPotato) [inline pkg/tso/allocator_manager.go:170]:
```suggestion
	// kgID is the keyspace group ID
```

File: pkg/tso/keyspace_group_manager.go [hunk 2]
```diff
@@ -55,20 +54,82 @@ const (
...
+			go func(am *AllocatorManager) {
+				defer wg.Done()
+				am.close()
+				log.Info("keyspace group closed", zap.Uint32("keyspace-group-id", am.kgID))
+			}(am)
+		}
+	}
+	wg.Wait()
+
+	log.Info("All keyspace groups closed")
```
Comment (JmPotato) [inline pkg/tso/keyspace_group_manager.go:92]:
```suggestion
	log.Info("all keyspace groups closed")
```

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: pkg/tso/keyspace_group_manager.go [hunk 3]
```diff
@@ -55,20 +54,82 @@ const (
...
+// getAMWithMembershipCheck returns the AllocatorManager of the given keyspace group and check
+// if the keyspace is served by this keyspace group.
+func (s *state) getAMWithMembershipCheck(
+	keyspaceID, keyspaceGroupID uint32,
+) (*AllocatorManager, uint32, error) {
+	s.RLock()
+	defer s.RUnlock()
+
+	if am := s.ams[keyspaceGroupID]; am != nil {
+		ksg := s.kgs[keyspaceGroupID]
```
[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:111]: changed.

File: pkg/tso/keyspace_group_manager.go [hunk 4]
```diff
@@ -439,22 +481,19 @@ func (kgm *KeyspaceGroupManager) isAssignedToMe(group *endpoint.KeyspaceGroup) b
...
-			zap.Uint32("keyspace-group-id", group.ID), zap.Int("max-keyspace-group-id", len(kgm.ams)-1))
+	if err := kgm.checkKeySpaceGroupID(group.ID); err != nil {
+		log.Warn("keyspace group ID is invalid, ignore it", zap.Error(err))
 		return
 	}
 
 	assignedToMe := kgm.isAssignedToMe(group)
 	if assignedToMe {
-		if kgm.ams[group.ID].Load() != nil {
+		if kgm.ams[group.ID] != nil {
```
[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:491]: No need. 

There are three writers -- initialization (Load the initial keyspace group assignment from storage), KeyspaceGroupsMetaWatchLoop (the code shown above is in this single goroutine), and shutdown. They are serialized, i.e., there are no concurrent writers. Specifically, we start KeyspaceGroupsMetaWatchLoop after static initialization, and the system exits from KeyspaceGroupsMetaWatchLoop before write access in shutdown.

In summary, KeyspaceGroupsMetaWatchLoop can read without lock, but it must require lock for write so that keyspace group manager can handle concurrent reads when processing requests.

File: pkg/tso/keyspace_group_manager.go [hunk 5]
```diff
@@ -519,53 +566,49 @@ func (kgm *KeyspaceGroupManager) updateKeyspaceGroupMembership(
...
-		kgm.keyspaceLookupTable.Store(kid, groupID)
-	}
-	return keyspaceLookupTable
+	kgm.kgs[groupID] = newGroup
 }
 
 // deleteKeyspaceGroup deletes the given keyspace group.
 func (kgm *KeyspaceGroupManager) deleteKeyspaceGroup(groupID uint32) {
-	kg := kgm.kgs[groupID].Swap(nil)
+	kg := kgm.kgs[groupID]
```
[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:598]: good catch. You saved me. I remember I added the lock here, but somehow it's gone.

[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:598]: kgm.Lock()
	defer kgm.Unlock()

File: pkg/tso/allocator_manager.go (hunk 1 above)
Comment (binshi-bing) [reply pkg/tso/allocator_manager.go:170]: changed.

File: pkg/tso/keyspace_group_manager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:598]: Confirm that no other places need lock

File: pkg/tso/keyspace_group_manager.go [hunk 6]
```diff
@@ -415,6 +462,8 @@ func (kgm *KeyspaceGroupManager) watchKeyspaceGroupsMetaChange(revision int64) (
 				case clientv3.EventTypeDelete:
 					kgm.deleteKeyspaceGroup(id)
 				}
+
+				kgm.storageChangeEventsApplied.Add(1)
```
Comment (JmPotato) [inline pkg/tso/keyspace_group_manager.go:466]: Could the counter overflow? Do you think we can only add it during the test? Maybe a failpoint could help.

File: pkg/tso/keyspace_group_manager.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rleungx) [reply pkg/tso/keyspace_group_manager.go:492]: But I suggest adding one to prevent the potential problem, it doesn't cost too much I think.

File: pkg/tso/keyspace_group_manager.go (hunk 6 above)
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go]: I already changed before saw your comment. Basically, I required read lock in the test. Check the commit 5.

File: pkg/tso/keyspace_group_manager.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:492]: It has cost. If I add a lock it needs to be a write lock and adding here will increase the length of the critical section.

[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:492]: The lock isn't re-entrant. If I'm right, with sync.RWMutex, I even can't require a read lock then upgrade it the write lock.

**Review by JmPotato — APPROVED** [review]

File: pkg/tso/keyspace_group_manager.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (binshi-bing) [reply pkg/tso/keyspace_group_manager.go:492]: good suggestion, actually.

changed to (kgm.getAllocatorManager requires the read lock)
```
		if oldAM, oldGroup := kgm.getAllocatorManager(group.ID); oldAM != nil {
			log.Info("keyspace group already initialized, so update meta only",
				zap.Uint32("keyspace-group-id", group.ID))
			kgm.updateKeyspaceGroupMembership(oldGroup, group)
			return
		}
```
